File: bench/histogram_bins.c
// benchmark-contract: histogram-xs13;seed=123456789;iterations=20000000;bins=64;index=state&63;xorshift=13
#include <stdint.h>
#include <stdio.h>

static inline void emit_checksum(uint64_t value) {
  // Every peer prints this one line and nothing else; the 63-bit mask
  // keeps the value printable by the languages without unsigned 64-bit
  // integers (NURL's `i`, Python's signed int, JS BigInt) so the five
  // outputs can be compared byte for byte.
  printf("%llu\n", (unsigned long long)(value & 0x7fffffffffffffffULL));
}

int main(void) {
  const uint64_t iterations = 20000000ULL;
  uint64_t state = 123456789ULL;
  uint64_t bins[64] = {0};

  uint64_t i = 0;
  while (i < iterations) {
    state = (state * 1664525ULL + 1013904223ULL) & 0xffffffffULL;
    state ^= state >> 13;
    uint64_t index = state & 63ULL;
    bins[index] += 1ULL;
    ++i;
  }

  uint64_t checksum = state;
  uint64_t index = 0;
  while (index < 64ULL) {
    checksum ^= bins[index] * (index + 1ULL);
    ++index;
  }
  emit_checksum(checksum);
  return 0;
}

File: bench/histogram_bins.rs
// benchmark-contract: histogram;seed=123456789;iterations=20000000;bins=64;index=state&63
const ITERATIONS: u64 = 20_000_000;
const BINS: usize = 64;
const SEED: u64 = 123_456_789;
const MASK: u64 = 0xffff_ffff;

fn finish(value: u64) -> ! {
    #[cfg(bench_verify)]
    std::io::Write::write_all(&mut std::io::stdout(), &value.to_le_bytes()).unwrap();
    std::process::exit((value & 0x7f) as i32)
}

fn main() {
    let mut state = SEED;
    let mut bins = [0u64; BINS];

    let mut i = 0u64;
    while i < ITERATIONS {
        state = state
            .wrapping_mul(1_664_525)
            .wrapping_add(1_013_904_223)
            & MASK;
        let index = (state & 63) as usize;
        bins[index] = bins[index].wrapping_add(1);
        i += 1;
    }

    let mut checksum = state;
    let mut index = 0usize;
    while index < BINS {
        checksum ^= bins[index].wrapping_mul(index as u64 + 1);
        index += 1;
    }

    finish(checksum);
}
